### DISCUSSION TITLE
get_dft_array frequencies, get total number of time steps

### DISCUSSION THREAD

**Discussion by dalarev** [NanoComp/meep#2336] [Q&A] [answered]
Hi,

Is there a way to grab the number of time steps performed in a simulation? This is not always known ahead of time. Sometimes these appear in the simulation report, but it would be useful to be able to grab them.

Is there a way to grab the frequencies at which `get_dft_array` is evaluated? 

These feature can be useful in their own right, but currently I'm needing to grab the total time steps because I think the `get_dft_array` fields need to be scaled by `1/N_time_steps`. I've seen before, but can't find at the moment, documentation that states these fields scale linearly with time.

Thanks.

**Comment by smartalecH** [comment]
You're probably thinking of the use of DFTs on a CW source. There's a few nuances to this approach, like normalizing by the number of time steps, since a CW source is not localized in time (extends forever). 

If you use a band-limited source (eg a Gaussian pulse) then the energy is finite as `t` approaches infinity. So if you divide by the number of time steps, the solution will approach 0...which is surely not what you want. In contrast, the dft here converges to a *finite value* with time. 

You can see details about how the dft is normalized within the meep paper. 

But generally, we just take advantage of linearity and normalize the fields with results from another run (eg in vacuum) such that any normalization constants cancel.

**Reply by dalarev** [reply]
Thanks for the notes. I had noticed that, even when using a Gaussian pulse, my results would scale with time. But perhaps this just means that my solutions weren't yet converged. Running more tests now.

Thanks very much!

**Comment by oskooi** [comment, marked as answer]
> Is there a way to grab the number of time steps performed in a simulation?

The number of elapsed timesteps in a `Simulation` class object is `Simulation.fields.t` (an integer). #2337 provides a wrapper function `timesteps()` for this object.

> Is there a way to grab the frequencies at which get_dft_array is evaluated?

Use the [`meep.get_flux_freqs`](https://meep.readthedocs.io/en/latest/Python_User_Interface/#flux-spectra) function.

**Reply by dalarev** [reply]
Works like a charm, thanks very much!
